coderd/database/migrations/000371_api_key_scopes_array_allow_list.up.sql

```diff
@@ -141,17 +141,25 @@ ALTER TYPE api_key_scope ADD VALUE IF NOT EXISTS 'workspace_proxy:read';
 ALTER TYPE api_key_scope ADD VALUE IF NOT EXISTS 'workspace_proxy:update';
 -- End enum extensions
 
+-- Purge old API keys to speed up the migration for large deployments.
+-- Note: that problem should be solved in coderd once PR 20863 is released:
+-- https://github.com/coder/coder/blob/main/coderd/database/dbpurge/dbpurge.go#L85
+DELETE FROM api_keys WHERE expires_at < NOW() - INTERVAL '7 days';
+
+--
 -- Add new columns without defaults; backfill; then enforce NOT NULL
-ALTER TABLE api_keys ADD COLUMN scopes api_key_scope[];
-ALTER TABLE api_keys ADD COLUMN allow_list text[];
+--
+ALTER TABLE api_keys
+    ADD COLUMN scopes api_key_scope[],
+    ADD COLUMN allow_list text[];
 
 -- Backfill existing rows for compatibility
-UPDATE api_keys SET scopes = ARRAY[scope::api_key_scope];
-UPDATE api_keys SET allow_list = ARRAY['*:*'];
-
--- Enforce NOT NULL
-ALTER TABLE api_keys ALTER COLUMN scopes SET NOT NULL;
-ALTER TABLE api_keys ALTER COLUMN allow_list SET NOT NULL;
+UPDATE api_keys SET
+    scopes = ARRAY[scope::api_key_scope],
+    allow_list = ARRAY['*:*'];
 
--- Drop legacy single-scope column
-ALTER TABLE api_keys DROP COLUMN scope;
+-- Enforce NOT NULL and drop legacy single-scope column
+ALTER TABLE api_keys
+    ALTER COLUMN scopes SET NOT NULL,
+    ALTER COLUMN allow_list SET NOT NULL,
+    DROP COLUMN scope;
```

Comment:
Just to keep the diff small, can we retain the old formatting where possible?

Reply:
How's that? I still kept the 2 `UPDATE`s rolled into one to avoid one extra scan (seems to save a few seconds on 500k rows). But I can undo that change too if those savings are not worth it.

Reply:
Nah keep it. My bad! I will review today